codex/stores/blockstore.nim

```diff
@@ -23,41 +24,39 @@ type
   OnBlock* = proc(cid: Cid): Future[void] {.upraises: [], gcsafe.}
   BlockStore* = ref object of RootObj
 
-method getBlock*(
-  b: BlockStore,
-  cid: Cid): Future[?!Block] {.base.} =
-  ## Get a block from the stores
+method getBlock*(self: BlockStore, cid: Cid): Future[?! (? Block)] {.base.} =
```

Comment:
This comes up in nim-datastore as well. My reasoning in that context was that an error reading from the database or filesystem, etc. is represented by the outer `?!`. If there was no read error as such but the block doesn't exist in the store then that's signaled by `none` for the value wrapped in `Option`.

If a missing block is an error in the same way as e.g. an OS error reading from the backend, it becomes a bit harder to distinguish those situations, especially if we're using `failure "some string"` as opposed to using a enum or an error type hierarchy to encode different errors.

Reply:
Here's the relevant code re: `SQLiteDatastore`:

https://github.com/status-im/nim-datastore/blob/96695fed47d44033145f947547a3b1c9958b3400/datastore/sqlite_datastore.nim#L293-L316

Reply:
> If a missing block is an error in the same way as e.g. an OS error reading from the backend, it becomes a bit harder to distinguish those situations, especially if we're using failure "some string" as opposed to using a enum or an error type hierarchy to encode different errors.

I more or less answered it in the top level comment -

> I'm not a big fan of the nested ?! (?...). It complicates usage and it almost never makes sense. Specially in the context of getBlock. IMO the distinction of wether the block is missing or we failed to retrieve it is moot and specially so in the context of a high level api, such as the block store.

Reply:
Sorry, I missed that in the top-level comment.

Do you want me to rework `method get` in nim-datastore? If so, I can make that change quickly.

Another option would be in nim-codex to wrap around the `?! (?...)` nesting in nim-datastore so that it's not exposed by nim-codex's API.

Reply:
Well, the API should be discussed in #123. This API is already implemented, and I think that our current goal is to include SQLite backend ASAP. So I propose to accept PR with this API, and make a new PR if we will decide to change it.

BTW, I copied API from Nim-Datastore because I considered it great design (and added Futures on top of it because you wanted to future-proof them :) ). But of course, when you want to handle everything explicitly instead of using exceptions and green threads, you get this sort of matryoshka datatypes :)

Reply:
The only thing I would insist we change is the `getBlock` return type, from `?! (?...)` to simply `?!`. I think this is a minor change. I'm OK if we change this in a separate PR as long as we do.

I don't think there is much we can do with `async` annotations and at this point I'd rather we have a nice clean api than a "super optimized" one.

Reply:
No, I don't agree that it's minor change. Situations 1) when we don't have the block requested and 2) when blockstore encountered error, are different. The question how we should deal with errors in blockstore is open. So I don't want to modify it only to modify it again after the next discussion.

My approach is to merge what is already implemented, discuss what we want to implement instead and then go to implement that.

Reply:
There are two issues here:

1) `Result[Option[Block], CatchableError]` violates [POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), the more appropriate way of communicating multiple error conditions with results is through an enum.
2) Is there anything useful we can do with the knowledge of the block missing vs failed from the standpoint of the caller? My argument is that there isn't


Like I said, I'm fine merging this PR as long as we address `getBlock` asap, even if it's in another PR.

Reply:
By address I mean that we figure out what is the best api for `getBlock`, I don't want to say that I'm 100% certain on being correct here 😉